c++/src/kj/compat/http.h

```diff
@@ -769,21 +769,47 @@ struct HttpServerSettings {
   // completes, we'll let the connection stay open to handle more requests.
 };
 
-class HttpServer final: private kj::TaskSet::ErrorHandler {
+class HttpServerErrorHandler {
+public:
+  virtual kj::Promise<void> requestError(
+      uint statusCode, kj::StringPtr statusText, kj::String errorMessage,
+      HttpHeaderTable& headerTable, kj::Maybe<HttpService::Response&> response) = 0;
+  // Similar to `HttpService::request()`, but called when the HttpServer encounters any sort of
+  // error, be it a protocol error or an exception thrown by the HttpService. The parameters
+  // `statusCode`, `statusText`, and `errorMessage`, are hints from the HttpServer suggesting what
+  // sort of response it thinks we should send to the client. The default implementation (which you
+  // must explicitly opt into by calling `HttpServerErrorHandler::requestError()` from your derived
+  // class) sends just such a response.
+  //
+  // If the HttpService did not already send a partial response, the `response` parameter will be
+  // non-null and available for you to customize the response sent to the client. You could use this
+  // to perform error message sanitization, e.g. to censor stack traces.
+  //
+  // Unlike `HttpService::request()`, when calling `response.send()` in this context, a
+  // "Connection: close" header will be added, and the connection will be closed.
+  //
+  // TODO(cleanup): It might make more sense to pass in the original exception object. But this
```

Comment:
I think we should pass in the exception object, so that the handler can inspect it and decide how to log it. E.g. maybe it wants to treat FAILED, OVERLOADED, and DISCONNECTED differently. Maybe the interface should have multiple methods:

    kj::Promise<void> handleClientProtocolError(kj::StringPtr message, ...);
    kj::Promise<void> handleApplicationError(kj::Exception e, ...);

(Are all the client protocol errors 400's, or are various status codes used?)

Reply:
Oh yeah, and if people want to censor stack traces then they definitely want to receive the `Exception` pre-stringification...

Reply:
I think there's just the one protocol error (bad headers), and it's a 400. In that light, the two function approach sounds great.